execution-engine/src/native_modules/darknet_inference.rs

```diff
@@ -0,0 +1,202 @@
+
+        // Prepare module's directory
+        create_dir(MODULE_DIR.as_path()).map_err(|_| ErrNo::Access)?;
+        set_current_dir(MODULE_DIR.as_path())?;
+
+        let result = self.infer(fs);
+        // NOTE: erase all the states and files passed to the module and created
+        // during execution
+        self.reset();
+        result
+    }
+
+    /// For the purpose of demonstration, we always return true. In reality,
+    /// this function may check validity of the `input`, and even buffer the
+    /// result for further uses.
+    fn try_parse(&mut self, input: &[u8]) -> FileSystemResult<bool> {
+        let deserialized_input: DarknetInferenceService =
+            match postcard::from_bytes(&input).map_err(|_| ErrNo::Canceled) {
+                Ok(o) => o,
+                Err(_) => return Ok(false),
+            };
+        *self = deserialized_input;
+        Ok(true)
+    }
+}
+
+impl DarknetInferenceService {
+    /// Create a new service, with empty internal state.
+    pub fn new() -> Self {
+        Self {
+            input_path: PathBuf::new(),
+            cfg_path: PathBuf::new(),
+            model_path: PathBuf::new(),
+            labels_path: PathBuf::new(),
+            output_path: PathBuf::new(),
+        }
+    }
+
+    /// The core service. It loads the model pointed by `model_path` with the
+    /// configuration in `cfg_path` and the labels defined in `labels_path`,
+    /// then feeds the input read from `input_path` to the model, and writes the
+    /// result to the file at `output_path`.
+    fn infer(&mut self, fs: &mut FileSystem) -> FileSystemResult<()> {
+        let DarknetInferenceService {
+            input_path,
+            cfg_path,
+            model_path,
+            labels_path,
+            output_path,
+        } = self;
+
+        let input_file_paths: [&PathBuf; 4] = [input_path, cfg_path, labels_path, model_path];
+
+        // Copy input files from the VFS to the kernel filesystem, preserving
+        // the file tree. At that point, we are already in the module's
+        // directory
+        for file_path in input_file_paths {
+            let parent_path = file_path.as_path().parent().ok_or(ErrNo::NoEnt)?;
+            let _ = create_dir_all(parent_path);
+            let buffer = fs.read_file_by_absolute_path(Path::new("/").join(file_path))?;
+            let mut file = File::create(file_path)?;
+            file.write_all(&buffer)?;
+        }
+
+        // Load network and labels
+        let mut net =
+            Network::load(cfg_path, Some(model_path), false).map_err(|_| ErrNo::Canceled)?;
+        let object_labels = read_to_string(labels_path)?
+            .lines()
+            .map(ToOwned::to_owned)
+            .collect::<Vec<_>>();
+
+        // Run inference
+        let image = Image::open(input_path).map_err(|_| ErrNo::Canceled)?;
+        let detections = net.predict(&image, 0.25, 0.5, 0.45, true);
```

Comment:
Where do these magic numbers come from?

Reply:
These are parameters (thresholds) to filter detections (after inference IIUC).
They should be passed as arguments to the module

Reply:
Alternatively we could process detections on the client side: unless the labels list is confidential, there's no real benefit in doing that in the enclave. That would also simplify the module's API and make it more generic